subsys/bluetooth/shell/bt.c

```diff
@@ -193,65 +193,128 @@ static bool data_cb(struct bt_data *data, void *user_data)
 	}
 }
 
-static void print_data_set(uint8_t type, uint8_t set_value_len,
+static void print_data_hex(const uint8_t *data, uint8_t len, bool warning)
+{
+	shell_fprintf(ctx_shell, warning ? SHELL_WARNING : SHELL_INFO, "0x");
+	/* Reverse the byte order when printing as advertising data is LE
+	 * and the MSB should be first in the printed output.
+	 */
+	for (int8_t i = len - 1; i >= 0; i--) {
+		shell_fprintf(ctx_shell, warning ? SHELL_WARNING : SHELL_INFO, "%02x", data[i]);
+	}
+}
+
+static void print_data_set(uint8_t set_value_len,
 			   const uint8_t *scan_data, uint8_t scan_data_len)
 {
-	uint8_t min_value_len = MIN(set_value_len, scan_data_len);
+	uint8_t idx = 0;
 
-	shell_fprintf(ctx_shell, SHELL_INFO, "%*sType 0x%02x: ",
-		      strlen(scan_response_label), "",
-		      type);
-	for (uint8_t i = 0U; i < (scan_data_len / min_value_len); i++) {
-		if (i > 0L) {
-			shell_fprintf(ctx_shell, SHELL_INFO, ", ");
-		}
+	if (scan_data_len == 0)
+		return;
 
-		shell_fprintf(ctx_shell, SHELL_INFO, "0x");
-		for (uint8_t j = 0U; j < min_value_len; j++) {
-			shell_fprintf(ctx_shell, SHELL_INFO, "%02x",
-				      scan_data[i * min_value_len + j]);
+	do {
+		if (idx > 0) {
+			shell_fprintf(ctx_shell, SHELL_INFO, ",");
 		}
-	}
+		print_data_hex(&scan_data[idx], set_value_len, false);
+		idx += set_value_len;
+	} while (idx + set_value_len <= scan_data_len);
 
-	shell_fprintf(ctx_shell, SHELL_INFO, "\n");
+	if (idx < scan_data_len) {
+		shell_fprintf(ctx_shell, SHELL_WARNING, " Excess data: ");
+		do {
+			print_data_hex(&scan_data[idx], 1, true);
+		} while (idx < scan_data_len);
+	}
 }
 
 static bool data_verbose_cb(struct bt_data *data, void *user_data)
 {
+	shell_fprintf(ctx_shell, SHELL_INFO, "%*sType 0x%02x: ",
+				  strlen(scan_response_label), "", data->type);
+
 	switch (data->type) {
 	case BT_DATA_UUID16_SOME:
 	case BT_DATA_UUID16_ALL:
 	case BT_DATA_SOLICIT16:
+		print_data_set(2, data->data, data->data_len);
+		break;
 	case BT_DATA_SVC_DATA16:
-		print_data_set(data->type, 2, data->data, data->data_len);
+		/* Data starts with a UUID16 (2 bytes),
+		 * the rest is unknown and printed as single bytes
+		 */
+		if (data->data_len < 2) {
+			shell_fprintf(ctx_shell, SHELL_WARNING,
+						  "BT_DATA_SVC_DATA16 data length too short (%u)",
+						  data->data_len);
+			break;
+		}
+		print_data_set(2, data->data, 2);
```

Comment:
If you aren't keen on using the `uuid_to_str`, I suggest to at least use the `BT_UUID_SIZE_XX` when parsing the UUIDs